### PR TITLE
journald event logging

### PR DESCRIPTION
**Pull request by baude**
add the ability for podman to read and write events to journald instead
of just a logfile.  This can be controlled in libpod.conf with the
`events_logger` attribute of `journald` or `file`.  The default will be
set to `journald`.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: cmd/podman/events.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ func eventsCmd(c *cliconfig.EventValues) error {
 	if err != nil {
 		return errors.Wrapf(err, "error creating libpod runtime")
 	}
+
```
Comment (rhatdan) [inline cmd/podman/events.go]: Remove this so the events.go does not get changed.

File: libpod/events.go [hunk 2]
```diff
@@ -50,51 +57,15 @@ func (v *Volume) newVolumeEvent(status events.Status) {
 	e := events.NewEvent(status)
 	e.Name = v.Name()
 	e.Type = events.Volume
-	if err := e.Write(v.runtime.config.EventsLogFilePath); err != nil {
-		logrus.Errorf("unable to write event to %s", v.runtime.config.EventsLogFilePath)
+	eventer := v.runtime.newEventer(&e)
+	if err := eventer.Write(); err != nil {
+		logrus.Errorf("unable to write container event")
```
Comment (rhatdan) [inline libpod/events.go]: Include the err.

File: libpod/events.go [hunk 3]
```diff
@@ -40,8 +46,9 @@ func (p *Pod) newPodEvent(status events.Status) {
 	e.ID = p.ID()
 	e.Name = p.Name()
 	e.Type = events.Pod
-	if err := e.Write(p.runtime.config.EventsLogFilePath); err != nil {
-		logrus.Errorf("unable to write event to %s", p.runtime.config.EventsLogFilePath)
+	eventer := p.runtime.newEventer(&e)
+	if err := eventer.Write(); err != nil {
+		logrus.Errorf("unable to write container event")
```
Comment (rhatdan) [inline libpod/events.go]: Include the err.

File: libpod/events/events.go [hunk 4]
```diff
@@ -4,12 +4,31 @@ import (
...
+type EventerType int
+
+const (
+	LogFile  EventerType = iota
+	Journald EventerType = iota
+)
+
+// String returns a string representation of EventerType
+func (et EventerType) String() string {
+	if et == Journald {
```
Comment (rhatdan) [inline libpod/events/events.go]: Reverse this
if et == File {
   return File
}
return Journald

File: libpod/events/events.go [hunk 5]
```diff
@@ -29,6 +48,230 @@ type Event struct {
...
+	// Until reads "until" the given time
+	Until string
+}
+
+// NewEventer creates an eventer based on the eventer type
+func NewEventer(event *Event, options EventerOptions) Eventer {
+	var eventer Eventer
+	switch strings.ToUpper(options.EventerType) {
+	case strings.ToUpper(Journald.String()):
+		eventer = EventJournalD{event, options}
```
Comment (rhatdan) [inline libpod/events/events.go]: Default to Journald

File: libpod/events/events.go [hunk 6]
```diff
@@ -29,6 +48,230 @@ type Event struct {
...
+// Read reads events from the journal and sends qualified events to the event channel
+func (e EventJournalD) Read(options ReadOptions) error {
+	eventOptions, err := generateEventOptions(options.Filters, options.Since, options.Until)
+	if err != nil {
+		return errors.Wrapf(err, "unable to generate event options")
+	}
+	podmanJournal := sdjournal.Match{Field: "SYSLOG_IDENTIFIER", Value: "podman"}
+	j, err := sdjournal.NewJournal()
+	if err != nil {
+		return err
```
Comment (rhatdan) [inline libpod/events/events.go]: Wrapf?

File: libpod/events/events.go [hunk 7]
```diff
@@ -29,6 +48,230 @@ type Event struct {
...
+	}
+	if err := j.AddMatch(podmanJournal.String()); err != nil {
+		return err
```
Comment (rhatdan) [inline libpod/events/events.go]: Wrapf?

File: libpod/events/events.go [hunk 8]
```diff
@@ -29,6 +48,230 @@ type Event struct {
...
+	}
+	if len(options.Since) == 0 && len(options.Until) == 0 {
+		if err := j.SeekTail(); err != nil {
+			return err
```
Comment (rhatdan) [inline libpod/events/events.go]: Wrapf?

File: libpod/events/events.go [hunk 9]
```diff
@@ -29,6 +49,96 @@ type Event struct {
...
+}
+
+// EventLogFile ...
+type EventLogFile struct {
+	*Event
+	options EventerOptions
+}
+
+// Writes to the log file
+func (e EventLogFile) Write() error {
```
[the remark this answers is not in the document]
Comment (mheon) [reply libpod/events/events.go]: Also - I think we dropped the write lock here

File: libpod/events/events_linux.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import "strings"
+
+// NewEventer creates an eventer based on the eventer type
+func NewEventer(event *Event, options EventerOptions) Eventer {
+	var eventer Eventer
+	switch strings.ToUpper(options.EventerType) {
+	case strings.ToUpper(LogFile.String()):
+		eventer = EventJournalD{event, options}
+	default:
```
Comment (mheon) [inline libpod/events/events_linux.go]: Don't make this default. Default should return an error.

File: libpod/events/filters.go [hunk 11]
```diff
@@ -1,20 +1,19 @@
-package shared
+package events
```
Comment (mheon) [inline libpod/events/filters.go]: Why did you need to move this - circular import issue?

File: libpod/events/events.go [hunk 12]
```diff
@@ -29,6 +49,96 @@ type Event struct {
...
+	// Read an event from the backend
+	Read(options ReadOptions) error
+}
+
+// ReadOptions describe the attributes needed to read event logs
+type ReadOptions struct {
+	// EventChannel is the comm path back to user
+	EventChannel chan *Event
+	// Filters are the results from the cli for --filter
+	Filters []string
```
Comment (mheon) [inline libpod/events/events.go]: This should be functions - look at how we filter images and containers

Comment (mheon) [reply libpod/events/events.go]: ...Can we not do that because journald? Damn it...

File: libpod/events/events.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (baude) [reply libpod/events/events.go]: i like the idea to split

File: libpod/events/journal_linux.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+		}
+		newCursor, err := j.GetCursor()
+		if err != nil {
+			return err
+		}
+		if prevCursor == newCursor {
+			if len(options.Until) > 0 || !options.Stream {
+				break
+			}
+			_ = j.Wait(sdjournal.IndefiniteWait) //nolint
```
[the remark this answers is not in the document]
Comment (baude) [reply libpod/events/journal_linux.go]: not an error, an int

File: libpod/events/events.go [hunk 14]
```diff
@@ -29,6 +49,96 @@ type Event struct {
...
+	// Write an event to a backend
+	Write() error
+	// Read an event from the backend
+	Read(options ReadOptions) error
+}
+
+// ReadOptions describe the attributes needed to read event logs
+type ReadOptions struct {
+	// EventChannel is the comm path back to user
+	EventChannel chan *Event
```
[the remark this answers is not in the document]
Comment (baude) [reply libpod/events/events.go]: Im going to ask you to give on this one... is that ok?

[the remark this answers is not in the document]
Comment (mheon) [reply libpod/events/events.go]: I'm not firm on this, but is there a reason?

File: libpod/events/events.go (hunk 12 above)
Comment (baude) [reply libpod/events/events.go]: agreed to defer to do this right, i..e to a map

**Comment by baude** [comment]
ping folks to review or merge this

**Comment by rhatdan** [comment]
I will wait for @mheon OK.

File: libpod/events/config.go [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+	// History ...
+	History Status = "history"
+	// Import ...
+	Import Status = "import"
+	// Init ...
+	Init Status = "init"
+	// Kill ...
+	Kill Status = "kill"
+	// LoadFromArchive ...
+	LoadFromArchive Status = "status"
```
Comment (mheon) [inline libpod/events/config.go]: This one doesn't make any sense - "status" as the event for loading from an archive?

**Comment by mheon** [comment]
I got one small thing, LGTM once that's fixed

**Comment by baude** [comment]
fixed

**Comment by rhatdan** [comment]
LGTM, Lets merge when it passes tests.

**Comment by mheon** [comment]
Agree

**Comment by TomSweeneyRedHat** [comment]
LGTM assuming happy tests
